src/Http/Routing/test/UnitTests/Internal/MapActionExpressionTreeBuilderTest.cs

```diff
@@ -24,44 +24,185 @@ namespace Microsoft.AspNetCore.Routing.Internal
+
+            var requestDelegate = MapActionExpressionTreeBuilder.BuildRequestDelegate(@delegate);
+
+            await requestDelegate(httpContext);
+
+            Assert.Equal(originalRouteParam, httpContext.Items["value"] as int?);
+        }
 
-            void TestAction([FromRoute] int value)
+        public static IEnumerable<object[]> FromRouteOptionalResult
+        {
+            get
             {
-                deserializedRouteParam = value;
+                return new List<object[]>
+                {
+                    new object[] { (Action<HttpContext, int>)TestAction },
+                    new object[] { (Func<HttpContext, int, Task>)TaskTestAction },
+                    new object[] { (Func<HttpContext, int, ValueTask>)ValueTaskTestAction }
+                };
             }
+        }
+
+        private static void TestAction(HttpContext httpContext, [FromRoute] int value = 42)
+        {
+            StoreInput(httpContext, value);
+        }
 
+        private static Task TaskTestAction(HttpContext httpContext, [FromRoute] int value = 42)
+        {
+            StoreInput(httpContext, value);
+            return Task.CompletedTask;
+        }
+
+        private static ValueTask ValueTaskTestAction(HttpContext httpContext, [FromRoute] int value = 42)
+        {
+            StoreInput(httpContext, value);
+            return ValueTask.CompletedTask;
+        }
+
+        [Theory]
+        [MemberData(nameof(FromRouteOptionalResult))]
+        public async Task RequestDelegatePopulatesFromRouteOptionalParameter(Delegate @delegate)
+        {
             var httpContext = new DefaultHttpContext();
-            httpContext.Request.RouteValues[paramName] = originalRouteParam.ToString(NumberFormatInfo.InvariantInfo);
 
-            var requestDelegate = MapActionExpressionTreeBuilder.BuildRequestDelegate((Action<int>)TestAction);
+            var requestDelegate = MapActionExpressionTreeBuilder.BuildRequestDelegate(@delegate);
 
             await requestDelegate(httpContext);
 
-            Assert.Equal(originalRouteParam, deserializedRouteParam);
+            Assert.Equal(42, httpContext.Items["value"] as int?);
+        }
+
+        [Theory]
+        [MemberData(nameof(FromRouteOptionalResult))]
+        public async Task RequestDelegatePopulatesFromRouteOptionalParameterBasedOnParameterName(Delegate @delegate)
+        {
+            const string paramName = "value";
+            const int originalRouteParam = 420;
+
+            var httpContext = new DefaultHttpContext();
+            httpContext.Items.Add("expected", originalRouteParam);
+
+            httpContext.Request.RouteValues[paramName] = originalRouteParam.ToString(NumberFormatInfo.InvariantInfo);
+
+            var requestDelegate = MapActionExpressionTreeBuilder.BuildRequestDelegate(@delegate);
+
+            await requestDelegate(httpContext);
```

Comment:
Still loving the tests! And the general cleanup. I'm not quite sure what this test is doing though. Is this missing some asserts?

Reply:
Yeah I forgot the Aseert, I will add it.